src/apps/app_bulk_upload.erl

```diff
@@ -0,0 +1,173 @@
+%%% submits them to Arweave as separate transactions. Transactions are submitted via
+%%% the queue module.
+
+-define(MB, 1048576). % 1024 * 1024
+-define(BLOB_HASH_ALGO, sha256).
+
+%% @doc Starts a queue server, splits the given file into chunks, wraps the chunks as 
+%% Arweave transactions, and submits them to the queue. Returns the queue PID.
+upload(Wallet, Filename) when is_list(Filename) ->
+	{ok, Filecontents} = file:read_file(Filename),
+	upload(Wallet, Filecontents);
+upload(Wallet, Blob) when is_binary(Blob) ->
+	upload(whereis(http_entrypoint_node), Wallet, Blob).
+
+upload(Node, Wallet, Blob) ->
+	Queue = app_queue:start(Node, Wallet),
+	Hash = crypto:hash(?BLOB_HASH_ALGO, Blob),
+	upload_blob(Queue, Hash, Blob),
+	Queue.
+
+%% @doc Takes a binary blob and processes it chunk by chunk. Each chunk is converted into
+%% a transaction and put into the queue. Chunk size is 1MB.
+upload_blob(Queue, Hash, Blob) ->
+	BlobSize = byte_size(Blob),
+	ChunkNumber = BlobSize div ?MB + if BlobSize rem ?MB =/= 0 -> 1; true -> 0 end,
+	upload_blob(Queue, Hash, Blob, ChunkNumber, 1).
+
+upload_blob(Queue, Hash, Blob, ChunkNumber, ChunkPosition) ->
+	if byte_size(Blob) =< ?MB ->
+		app_queue:add(Queue, chunk_to_tx(Hash, Blob, ChunkNumber, ChunkPosition));
+	true ->
+		<< Chunk:?MB/binary, Rest/binary >> = Blob,
+		app_queue:add(Queue, chunk_to_tx(Hash, Chunk, ChunkNumber, ChunkPosition)),
+		upload_blob(Queue, Hash, Rest, ChunkNumber, ChunkPosition + 1)
+	end.
+
+%% @doc Converts the given binary chunk into a transaction. A hash of the whole block the chunk
+%% is part of is assigned as a tag.
+chunk_to_tx(Hash, Chunk, ChunkNumber, ChunkPosition) ->
+	#tx {
+		tags =
+			[
+				{"app_name", "BulkUpload"},
+				{"blob_hash", ar_util:encode(Hash)},
+				{"number_of_chunks", integer_to_binary(ChunkNumber)},
+				{"chunk_position", integer_to_binary(ChunkPosition)}
+			],
+		data = Chunk
+	}.
+
+%% @doc Searches the local storage for the chunks of the blob with the given hash.
+%% If the blob is reconstructed successfuly, writes it to the specified destination.
+%% The provided hash has to be a Base 64 encoded SHA 256 hash of the file as a binary string.
+download(Hash, Filename) ->
+	{ok, Blob} = download(Hash),
+	file:write_file(Filename, Blob, [{encoding, unicode}]).
+
+download(Hash) ->
+	app_search:get_entries(<< "blob_hash" >>, Hash),
+	receive TXIDs ->
+		Transactions = lists:map(fun(TX) -> ar_storage:read_tx(TX) end, TXIDs),
+		{ok, Blob} = reconstruct_blob(Transactions),
+		BlobHash = ar_util:encode(crypto:hash(?BLOB_HASH_ALGO, Blob)),
+		if BlobHash /= Hash ->
+			{error, invalid_upload};
+		true ->
+			{ok, Blob}
+		end
+	end.
+
+reconstruct_blob([]) -> {error, not_found};
+reconstruct_blob(Transactions) ->
+	Head = hd(Transactions),
+	{_, ChunkNumberBinary} = lists:keyfind(<< "number_of_chunks" >>, 1, Head#tx.tags),
+	ChunkNumber = binary_to_integer(ChunkNumberBinary),
+	if length(Transactions) < ChunkNumber ->
```

Comment:
If the same file is uploaded twice (by the same user or by different users) this will still pass. Although, it's not enough to change it to `if length(Transactions) /= ChunkNumber ->`.